File: src/model/nonParQuantileCausality-main/nonParQuantileCausality-main/R/np_quantile_causality.R
#' Nonparametric Causality-in-Quantiles Test
#'
#' @description
#' Computes the Balcilar-Jeong-Nishiyama style nonparametric quantile
#' Granger-causality test for first-order lags. Methodology is based on 
#' Balcilar, Gupta, and Pierdzioch (2016, \doi{10.1016/j.resourpol.2016.04.004})
#' and Balcilar et al. (2016, \doi{10.1007/s11079-016-9388-x}).
#'
#' @param x numeric vector; candidate cause (independent) variable. The test internally uses the
#'   **first lag** of x (one-lag Granger causality setup).
#' @param y numeric vector; effect (dependent) variable. The test internally uses the
#'   **first lag** of y (one-lag Granger causality setup).
#' @param type character; "mean" or "variance" (causality in mean or variance).
#' @param q numeric vector of quantiles in (0,1). Default is seq(0.01, 0.99, 0.01).
#' @param hm optional numeric bandwidth; if `NULL`, uses Yu & Jones (1998) style
#'   plug-in via \code{KernSmooth::dpill} on the mean-regression proxy.
#'
#' @return An object of class \code{np_quantile_causality} with elements:
#' \itemize{
#'   \item \code{statistic}: numeric vector of test statistics by quantile
#'   \item \code{quantiles}: numeric vector of quantiles tested
#'   \item \code{bandwidth}: scalar base bandwidth used before quantile adjustment
#'   \item \code{type}: "mean" or "variance"
#'   \item \code{n}: effective sample size
#'   \item \code{call}: the matched call
#' }
#'
#' @details
#' Uses local polynomial quantile regression at each quantile with kernel weights,
#' constructs the Song et al. (2012) style quadratic form, and rescales to the
#' asymptotic standard-normal statistic.
#'
#' @section Lag order (important):
#' The current implementation **uses one lag** of each series only:
#' \eqn{x_{t-1}} and \eqn{y_{t-1}} (first-order Granger setup).
#' Extending to higher lags requires changing the internal embedding
#' (currently `stats::embed(*, 2)`) and the kernel construction to handle
#' multivariate lag vectors (e.g., a product kernel over all lag coordinates
#' or a multivariate Gaussian kernel).
#'
#' @note This function tests whether \eqn{x_{t-1}} Granger-causes \eqn{y_t}
#' in quantile \eqn{\theta} (and, with `type = "variance"`, whether
#' \eqn{x_{t-1}^2} causes \eqn{y_t^2}). Higher-order lags are **not** supported
#' in this release.
#'
#' @examples
#' \donttest{
#' set.seed(1234)
#' x <- arima.sim(n = 600, list(ar = 0.4))
#' y <- 0.5*lag(x, -1) + rnorm(600)  # x Granger-causes y
#' y[is.na(y)] <- mean(y, na.rm = TRUE)
#' obj <- np_quantile_causality(x, y, type = "mean", q = seq(0.1, 0.9, 0.1))
#' plot(obj)  # test statistic vs quantiles with 5% CV line
#'
#' # Example with bundled dataset (Gold causes Gold or Oil depending on call)
#' data(gold_oil)
#' # use first 500 days
#' gold_oil <- gold_oil[1:501,]
#' q_grid <- seq(0.25, 0.75, by = 0.25)
#'
#' # Causality in conditional mean (does Oil_t-1 cause Gold_t?)
#' res_mean <- np_quantile_causality(
#'   x = gold_oil$Oil,
#'   y = gold_oil$Gold,
#'   type = "mean",
#'   q = q_grid
#' )
#' res_mean
#'
#' # Causality in conditional variance
#' res_var <- np_quantile_causality(
#'   x = gold_oil$Oil,
#'   y = gold_oil$Gold,
#'   type = "variance",
#'   q = q_grid
#' )
#' res_var
#'
#' # Plot (with 5% critical value line); returns a ggplot object invisibly
#' plot(res_mean)
#' plot(res_var)
#' }
#'
#' @section References:
#' \itemize{
#'   \item Balcilar, M., Gupta, R., & Pierdzioch, C. (2016).
#'         Does uncertainty move the gold price? New evidence from a nonparametric
#'         causality-in-quantiles test. \emph{Resources Policy}, 49, 74–80.
#'         \doi{10.1016/j.resourpol.2016.04.004}
#'   \item Balcilar, M., Gupta, R., Kyei, C., & Wohar, M. E. (2016).
#'         Does economic policy uncertainty predict exchange rate returns and volatility?
#'         Evidence from a nonparametric causality-in-quantiles test.
#'         \emph{Open Economies Review}, 27(2), 229–250.
#'         \doi{10.1007/s11079-016-9388-x}
#' }
#'
#' @export
np_quantile_causality <- function(x,
                                  y,
                                  type = c("mean", "variance"),
                                  q = NULL,
                                  hm = NULL) {
  cl <- match.call()
  type <- match.arg(type)
  
  if (!is.numeric(x) ||
      !is.numeric(y))
    stop("x and y must be numeric vectors.")
  if (length(x) != length(y))
    stop("x and y must have the same length.")
  if (is.null(q))
    qvec <- seq(0.01, 0.99, by = 0.01)
  else
    qvec <- q
  if (any(qvec <= 0 |
          qvec >= 1))
    stop("All quantiles must be in (0,1).")
  
  # First lag embedding
  if (length(y) < 3)
    stop("Need at least length >= 3.")
  y_all <- stats::embed(y, 2)
  y_lag1 <- y_all[, 2]  # y_{t-1}
  y_t    <- y_all[, 1]  # y_t
  
  x_all <- stats::embed(x, 2)
  x_lag1 <- x_all[, 2]  # x_{t-1}
  
  # Switch for mean/variance causality
  if (type == "variance") {
    y2 <- y^2
    x2 <- x^2
  } else {
    y2 <- y
    x2 <- x
  }
  
  y2_all <- stats::embed(y2, 2)
  y2_t   <- y2_all[, 1]   # y2_t
  y2_lag1 <- y2_all[, 2]  # y2_{t-1}
  
  # NOTE: the original code had x2_all built from y2; fix to use x2.
  x2_all <- stats::embed(x2, 2)
  x2_t    <- x2_all[, 1]   # x2_t (unused downstream but kept for symmetry)
  x2_lag1 <- x2_all[, 2]   # x2_{t-1}
  
  tn <- length(y_t)  # effective n
  
  # Base bandwidth from mean-regression proxy (Yu & Jones, 1998) via dpill
  if (is.null(hm)) {
    # Use y_{t-1} -> y_t mean-regression proxy for dpill inputs
    h_base <- KernSmooth::dpill(y_lag1, y_t, gridsize = tn)
  } else {
    h_base <- hm
  }
  
  tstat_vec <- numeric(length(qvec))
  
  # Loop over quantiles
  for (j in seq_along(qvec)) {
    qj <- qvec[j]
    # Quantile-specific bandwidth adjustment
    qrh <- h_base * ((qj * (1 - qj) / (stats::dnorm(
      stats::qnorm(qj)
    )^2))^(1 / 5))
    
    # Local linear quantile regression of y2_t on y2_{t-1}, evaluated at y_{t-1}
    fit <- lprq2_(
      x = y2_lag1,
      y = y2_t,
      h = qrh,
      tau = qj,
      x0 = y_lag1
    )
    
    # Indicator residuals I(y2_t <= \hat{Q}_tau) - tau
    if_temp <- as.numeric(y2_t <= fit$fv) - qj
    if_vec  <- matrix(if_temp, ncol = 1)
    
    # Kernel weight matrix K over (y_{t-1}, x_{t-1}) distances, scaled
    # Using Gaussian product kernel
    y_mat <- outer(y_lag1, y_lag1, "-")
    x_mat <- outer(x_lag1, x_lag1, "-")
    # Scale x kernel by relative sd to y (matches original code intent)
    scale_x <- (stats::sd(y_lag1) / stats::sd(x_lag1))
    K <- stats::dnorm(y_mat / qrh) * stats::dnorm((x_mat / qrh) * scale_x)
    
    num <- t(if_vec) %*% K %*% if_vec
    den <- sqrt(tn / (2 * qj * (1 - qj)) / (tn - 1) / sum(K^2))
    
    tstat_vec[j] <- as.numeric(num * den) # asymptotic N(0,1) under H0
  }
  
  out <- list(
    statistic = tstat_vec,
    quantiles = qvec,
    bandwidth = h_base,
    type = type,
    n = tn,
    call = cl
  )
  class(out) <- "np_quantile_causality"
  out
}

#' @export
#' @rdname np_quantile_causality
#' @usage NULL
lrq_causality_test <- function(...) {
  .Deprecated("np_quantile_causality",
              package = "nonParQuantileCausality",
              msg = "Use np_quantile_causality(); dots-in-names are deprecated.")
  np_quantile_causality(...)
}

File: src/model/np_quantile_causality.py
"""Nonparametric Causality-in-Quantiles Test — Python translation.

Faithful translation of the R package `nonParQuantileCausality`
(Balcilar-Jeong-Nishiyama style test, Jeong et al. 2012).

References:
    Jeong, K., Härdle, W.K., Song, S. (2012).
        A consistent nonparametric test for causality in quantile.
        Econometric Theory 28, 861–887.
    Balcilar, M., Gupta, R., Pierdzioch, C. (2016).
        Resources Policy, 49, 74–80.
"""

from typing import Dict, List, Optional, Tuple

import numpy as np
from scipy.stats import norm


# ---------------------------------------------------------------------------
# 1. dpill — Silverman plug-in bandwidth (substitute for KernSmooth::dpill)
# ---------------------------------------------------------------------------

def _dpill(x: np.ndarray, y: np.ndarray) -> float:
    """Plug-in bandwidth for local linear regression (Silverman's rule).

    R's KernSmooth::dpill uses the Sheather-Jones direct plug-in method.
    We approximate with Silverman's rule-of-thumb, which is adequate here
    because the Yu & Jones quantile adjustment dominates the final bandwidth.
    """
    n = len(x)
    std_x = np.std(x, ddof=1)
    iqr_x = np.subtract(*np.percentile(x, [75, 25]))
    spread = min(std_x, iqr_x / 1.34) if iqr_x > 0 else std_x
    if spread <= 0:
        spread = 1.0
    h = 0.9 * spread * n ** (-0.2)
    return h


# ---------------------------------------------------------------------------
# 2. Weighted quantile regression via IRLS (replaces LP formulation)
# ---------------------------------------------------------------------------

def _weighted_quantile_regression(
    y: np.ndarray,
    X: np.ndarray,
    weights: np.ndarray,
    tau: float,
    max_iter: int = 50,
    eps: float = 1e-6,
) -> np.ndarray:
    """Weighted quantile regression via IRLS (MM algorithm).

    Solves the same problem as the LP version:
        min  sum_i w_i * rho_tau(y_i - X_i @ beta)
    but uses iteratively reweighted least squares with a smoothed
    check function, avoiding the O(n²) LP constraint matrix.

    Args:
        y: response vector (n,)
        X: design matrix (n, p), should include intercept column
        weights: non-negative weights (n,)
        tau: quantile level in (0, 1)
        max_iter: maximum IRLS iterations
        eps: residual threshold to avoid division by zero

    Returns:
        beta: coefficient vector (p,)
    """
    n, p = X.shape
    w = np.maximum(weights, 0.0)

    # Initialize with weighted least squares (tau=0.5 approximation)
    W_diag = w + 1e-12
    XtW = X.T * W_diag
    try:
        beta = np.linalg.solve(XtW @ X + 1e-10 * np.eye(p), XtW @ y)
    except np.linalg.LinAlgError:
        return np.full(p, np.nan)

    for _ in range(max_iter):
        resid = y - X @ beta
        # IRLS weights: rho_tau'(u) / |u|, smoothed near zero
        abs_resid = np.maximum(np.abs(resid), eps)
        irls_w = w * np.where(resid >= 0, tau, 1.0 - tau) / abs_resid

        XtIW = X.T * irls_w
        try:
            beta_new = np.linalg.solve(XtIW @ X + 1e-10 * np.eye(p), XtIW @ y)
        except np.linalg.LinAlgError:
            break

        if np.max(np.abs(beta_new - beta)) < eps:
            return beta_new
        beta = beta_new

    return beta


# ---------------------------------------------------------------------------
# 3. lprq2_ — local polynomial quantile regression (R/utils_internal.R)
# ---------------------------------------------------------------------------

def _lprq2(
    x: np.ndarray,
    y: np.ndarray,
    h: float,
    tau: float,
    x0: np.ndarray,
) -> np.ndarray:
    """Local linear quantile regression evaluated at x0.

    For each evaluation point x0[i], fits a weighted quantile regression
        y ~ intercept + (x - x[i])
    with Gaussian kernel weights dnorm((x0 - x0[i]) / h).
    """
    n = len(x0)
    fv = np.empty(n)
    ones = np.ones(len(x))
    inv_h = 1.0 / h
    inv_sqrt_2pi = 1.0 / np.sqrt(2.0 * np.pi)
    X_design = np.empty((len(x), 2))
    X_design[:, 0] = 1.0

    for i in range(n):
        z = x - x[i]
        z0_scaled = (x0 - x0[i]) * inv_h
        w = inv_sqrt_2pi * np.exp(-0.5 * z0_scaled * z0_scaled)

        X_design[:, 1] = z
        beta = _weighted_quantile_regression(y, X_design, w, tau)
        fv[i] = beta[0]

    return fv


# ---------------------------------------------------------------------------
# 4. np_quantile_causality — main test (R/np_quantile_causality.R)
# ---------------------------------------------------------------------------

def np_quantile_causality(
    x: np.ndarray,
    y: np.ndarray,
    test_type: str = "mean",
    q: Optional[List[float]] = None,
    hm: Optional[float] = None,
) -> Dict:
    """Nonparametric causality-in-quantiles test (Jeong et al. 2012).

    Tests whether x_{t-1} Granger-causes y_t at quantile tau.
    Uses first-order lags only (lag = 1).

    Faithful translation of R function np_quantile_causality().

    Args:
        x: candidate cause series (T,)
        y: effect series (T,)
        test_type: "mean" (causality in conditional mean) or
                   "variance" (causality in conditional variance, uses x^2, y^2)
        q: list of quantiles in (0,1). Default: [0.01, 0.02, ..., 0.99]
        hm: optional base bandwidth. If None, uses Silverman plug-in.

    Returns:
        dict with keys:
            'statistic': array of test statistics (asymptotic N(0,1) under H0)
            'quantiles': array of quantile levels tested
            'bandwidth': base bandwidth used
            'type': "mean" or "variance"
            'n': effective sample size
    """
    x = np.asarray(x, dtype=float)
    y = np.asarray(y, dtype=float)

    if x.shape != y.shape or x.ndim != 1:
        raise ValueError("x and y must be 1-d numeric vectors of equal length.")
    if len(y) < 3:
        raise ValueError("Need at least length >= 3.")
    if test_type not in ("mean", "variance"):
        raise ValueError("type must be 'mean' or 'variance'.")

    if q is None:
        qvec = np.arange(0.01, 1.0, 0.01)
    else:
        qvec = np.asarray(q, dtype=float)
    if np.any(qvec <= 0) or np.any(qvec >= 1):
        raise ValueError("All quantiles must be in (0, 1).")

    # --- First lag embedding (R: embed(y, 2)) ---
    y_t = y[1:]         # y_all[, 1] in R (current)
    y_lag1 = y[:-1]     # y_all[, 2] in R (lag-1)
    x_lag1 = x[:-1]     # x_all[, 2] in R (lag-1)

    # --- Mean / variance switch ---
    if test_type == "variance":
        y2 = y ** 2
        x2 = x ** 2
    else:
        y2 = y.copy()
        x2 = x.copy()

    y2_t = y2[1:]
    y2_lag1 = y2[:-1]
    x2_lag1 = x2[:-1]  # noqa: F841 (kept for symmetry with R code)

    tn = len(y_t)  # effective sample size

    # --- Base bandwidth ---
    if hm is None:
        h_base = _dpill(y_lag1, y_t)
    else:
        h_base = float(hm)

    tstat_vec = np.empty(len(qvec))

    # --- Loop over quantiles ---
    for j, qj in enumerate(qvec):
        # Yu & Jones (1998) quantile-specific bandwidth adjustment
        qrh = h_base * ((qj * (1 - qj) / (norm.pdf(norm.ppf(qj)) ** 2)) ** 0.2)

        # Local linear quantile regression of y2_t on y2_{t-1}, at y_{t-1}
        fv = _lprq2(x=y2_lag1, y=y2_t, h=qrh, tau=qj, x0=y_lag1)

        # Indicator residuals: I(y2_t <= Q_hat) - tau
        if_vec = (y2_t <= fv).astype(float) - qj  # (tn,)

        # Kernel matrix K (tn x tn) — Gaussian product kernel
        y_mat = y_lag1[:, None] - y_lag1[None, :]      # R: outer(y_lag1, y_lag1, "-")
        x_mat = x_lag1[:, None] - x_lag1[None, :]      # R: outer(x_lag1, x_lag1, "-")
        scale_x = np.std(y_lag1, ddof=1) / np.std(x_lag1, ddof=1)
        K = norm.pdf(y_mat / qrh) * norm.pdf((x_mat / qrh) * scale_x)

        # Test statistic (Song et al. 2012 quadratic form)
        num = if_vec @ K @ if_vec                       # R: t(if_vec) %*% K %*% if_vec
        den = np.sqrt(tn / (2 * qj * (1 - qj)) / (tn - 1) / np.sum(K ** 2))
        tstat_vec[j] = float(num * den)                 # N(0,1) under H0

    return {
        "statistic": tstat_vec,
        "quantiles": qvec,
        "bandwidth": h_base,
        "type": test_type,
        "n": tn,
    }


# ---------------------------------------------------------------------------
# 5. qn_causality_network — build K×K adjacency matrix
# ---------------------------------------------------------------------------

def qn_causality_network(
    data: np.ndarray,
    tau: float = 0.05,
    alpha: float = 0.05,
    test_type: str = "mean",
) -> np.ndarray:
    """Build K×K directed adjacency matrix using QN causality test.

    For each ordered pair (j → i), tests whether series j Granger-causes
    series i at quantile tau using the Jeong et al. (2012) nonparametric test.

    Args:
        data: T×K array (e.g. risk measure for K industries)
        tau: quantile level for causality test (default 0.05 = 5% left tail)
        alpha: significance level for the test (default 0.05)
        test_type: "mean" or "variance"

    Returns:
        W: K×K binary adjacency matrix. W[i,j]=1 means j→i is significant.
    """
    T, K = data.shape
    W = np.zeros((K, K))
    cv = norm.ppf(1 - alpha)  # one-sided critical value

    for i in range(K):
        for j in range(K):
            if i == j:
                continue
            result = np_quantile_causality(
                x=data[:, j], y=data[:, i],
                test_type=test_type, q=[tau],
            )
            if result["statistic"][0] > cv:
                W[i, j] = 1.0

    return W


# ---------------------------------------------------------------------------
# 6. Validation / self-test
# ---------------------------------------------------------------------------

def _validate():
    """Run basic validation: synthetic data with known causality."""
    print("=" * 60)
    print("Validation: np_quantile_causality Python translation")
    print("=" * 60)

    np.random.seed(1234)

    # --- Test 1: x causes y ---
    n = 600
    x = np.empty(n)
    x[0] = np.random.randn()
    for t in range(1, n):
        x[t] = 0.4 * x[t - 1] + np.random.randn()
    y = np.empty(n)
    y[0] = np.random.randn()
    for t in range(1, n):
        y[t] = 0.5 * x[t - 1] + np.random.randn()

    q_grid = [0.05, 0.25, 0.50, 0.75, 0.95]
    result = np_quantile_causality(x, y, test_type="mean", q=q_grid)

    print("\nTest 1: x -> y (should reject H0, tstat > 1.645)")
    print(f"  bandwidth = {result['bandwidth']:.4f}")
    print(f"  n = {result['n']}")
    for qi, ts in zip(result["quantiles"], result["statistic"]):
        sig = " ***" if ts > 1.645 else ""
        print(f"  tau={qi:.2f}  tstat={ts:.4f}{sig}")

    # --- Test 2: independent series ---
    z = np.random.randn(n)
    w = np.random.randn(n)
    result2 = np_quantile_causality(z, w, test_type="mean", q=q_grid)

    print("\nTest 2: independent z, w (should NOT reject H0)")
    for qi, ts in zip(result2["quantiles"], result2["statistic"]):
        sig = " *** FALSE POSITIVE" if ts > 1.645 else ""
        print(f"  tau={qi:.2f}  tstat={ts:.4f}{sig}")

    # --- Test 3: network construction ---
    print("\nTest 3: 5-industry network (synthetic data)")
    K_test = 5
    T_test = 300
    data_net = np.random.randn(T_test, K_test)
    # inject causality: industry 0 -> industry 1
    for t in range(1, T_test):
        data_net[t, 1] += 0.5 * data_net[t - 1, 0]

    W = qn_causality_network(data_net, tau=0.50, alpha=0.05, test_type="mean")
    density = W.sum() / (K_test * (K_test - 1))
    print(f"  density = {density:.3f}")
    print(f"  W[1,0] (0->1 should be 1) = {W[1, 0]:.0f}")
    print(f"  W matrix:\n{W}")

    print("\n" + "=" * 60)
    print("Validation complete.")
    print("=" * 60)


if __name__ == "__main__":
    _validate()
